### PR TITLE
Fix Firefox Bugs

### PR DESCRIPTION
**Pull request by matthewborden**
Fix Firefox issues:
- Not being able to use arrow keys to navigate past equations.
- Fix backspace (again, firefox wasn't buying into any of the old hacks).

### CODE REVIEW

File: source/Editor.js [hunk 1]
```diff
@@ -1993,11 +1993,49 @@ var keyHandlers = {
             // If the commonAncestor was in a mathjax equation, delete the parent. Namely the mathjax equation.
             if (range.commonAncestorContainer && 
                 range.commonAncestorContainer.parentElement && 
-                checkInSVG.test(range.commonAncestorContainer.nodeName)
+                (checkInSVG.test(range.commonAncestorContainer.nodeName) || checkInSVG.test(range.commonAncestorContainer.parentElement.nodeName))
```
Comment (KJTsanaktsidis) [inline source/Editor.js]: Could this not be fixed just by adding more svg tags to the definition of `checkInSVG`, or by checking what kind of document commonAncestorContainer is in? What is this trying to achieve?

File: source/Editor.js [hunk 2]
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
         // Test conditions for selection being at the start of the textnode.
         var validNodes = !!(anchorNode && anchorPreviousNode);
         var caretAtStartOfNode = selection.anchorOffset === 0;
-
+        var isMathjaxIOS, isMathjaxIE, isMathjax;
```
Comment (KJTsanaktsidis) [inline source/Editor.js]: Why this change?

File: source/Editor.js [hunk 3]
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
...
             // this is a mathjax equation, prevent defualt.
             event.preventDefault();
 
             // create a new range
-            var leftRange = self._doc.createRange();
             leftRange.setStartBefore(anchorPreviousNode);
             leftRange.setEndBefore(anchorPreviousNode);
-            
+
+            if (isGecko && anchorPreviousNode.previousSibling && anchorPreviousNode.previousSibling.textContent) {
```
Comment (KJTsanaktsidis) [inline source/Editor.js]: Checking the node type of `previousSibling` would convey your intent here more clearly

File: source/Editor.js [hunk 4]
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
...
+                var beforePreviousNode = anchorPreviousNode.previousSibling;
+                leftRange.setStart(beforePreviousNode, beforePreviousNode.textContent.length);
+                leftRange.setEnd(beforePreviousNode, beforePreviousNode.textContent.length);
+            }
+
             // set a selection.
             self.setSelection(leftRange);
+            return;
```
Comment (KJTsanaktsidis) [inline source/Editor.js]: This looks pretty redundant

File: source/Editor.js (hunk 1 above)
Comment (matthewborden) [inline source/Editor.js]: This is trying to catch strange cases where text nodes are passed by the browser as commonAncestorContainer. The most common occurrence of this, is that the text node within a script tag will be passed as the commonAncestorContainer. This intends to confirm that we've covered all the bases, in terms of what the browser could send us.

Comment (KJTsanaktsidis) [inline source/Editor.js]: Are we talking about text nodes inside the SVG document?

Comment (matthewborden) [inline source/Editor.js]: In that case I was talking about a structure that looked like.
    SPAN.mathjax
        -> SCRIPT
             -> TEXT NODE (The browser passes this one).
 ^^ GH removes the indentation of the tree (edit the comment to view)

I don't think it's too offensive to confirm that the parent of what was passed wasn't an SVG element or a Script tag. Those things should be delete anyway shouldn't they?

Comment (KJTsanaktsidis) [inline source/Editor.js]: Is the script outside of the svg element? eew.

Comment (KJTsanaktsidis) [inline source/Editor.js]: In that case, isn't what you're really interested in whether or not range.commonAncestorContainer is a child of a `span.mathjax`?

Comment (matthewborden) [inline source/Editor.js]: Yep. Mathjax is great at generating disgusting things.

Comment (KJTsanaktsidis) [inline source/Editor.js]: It'd be good if you can make the condition of this if statement clearly express that; i.e. do something like the while loop below (terminating at the top of the body or something) to see if the commonAncestorContainer is a child of span.mathjax

Comment (matthewborden) [inline source/Editor.js]: 'top of the body' by this are you referring to the document body? Just wanted to confirm.

Comment (KJTsanaktsidis) [inline source/Editor.js]: The iframe body

Comment (matthewborden) [inline source/Editor.js]: Sorry, I just re-read you're earlier comments. Yes, I'm interested if the range.commonAncestorContainer is a child of a Mathjax equation. 

The loop terminates at the body now. You're advocating that the condition be something like:

```
if (mathjaxParent.className === "mathjax''){
     if(mathjaxParent.contains(range.commonAncestorContainer)){
           // remove the mathjaxParent.
     }
     break;
}
```

Comment (KJTsanaktsidis) [inline source/Editor.js]: More like

``` javascript
var currentSearchNode = range.commonAncestorContainer;
var mathjaxParent = null;
while (currentSearchNode !== self._doc.body) {
    if (currentSearchNode.classList.contains('mathjax')) {
        mathjaxParent = currentSearchNode;
        break;
    }
    currentSearchNode = currentSearchNode.parent;
}

if (mathjaxParent) {
    event.preventDefault();
    detach(mathjaxParent);
}
```

i.e. it looks like you want to prevent default if the common ancestor is in a span.mathjax.

Comment (KJTsanaktsidis) [inline source/Editor.js]: It kinda sucks the DOM api hasn't got a better way of doing this.

Comment (KJTsanaktsidis) [inline source/Editor.js]: The other way of expressing it might be

``` javascript
var mathjaxes = self._doc.querySelectorAll('span.mathjax');
var mathjaxParent = _.find(mathjaxes, function(mathjaxEl) {
    return mathjaxEl.contains(range.commonAncestorContainer);
});


if (mathjaxParent) {
    event.preventDefault();
    detach(mathjaxParent);
}
```

Excercise for the reader: do this without underscore :(

Comment (KJTsanaktsidis) [inline source/Editor.js]: @slashgrin likes the latter. I think I do too.
